src/Stratis.Bitcoin.Features.BlockStore/LoopSteps/ProcessPendingStorageStep.cs

```diff
@@ -52,95 +52,27 @@ internal ProcessPendingStorageStep(BlockStoreLoop blockStoreLoop, ILoggerFactory
-
-            while (this.BlockStoreLoop.PendingStorage.Count > 0)
+            
+            // In case of IBD do not save every single block- persist them in batches.
+            if (this.BlockStoreLoop.PendingStorage.Count < BlockStoreLoop.PendingStorageBatchThreshold &&
+                !disposeMode && this.BlockStoreLoop.InitialBlockDownloadState.IsInitialBlockDownload())
             {
-                StepResult result = this.PrepareNextBlockFromPendingStorage(context);
-                if (result == StepResult.Stop)
-                    break;
+                return StepResult.Stop;
             }
 
-            if (context.PendingBlockPairsToStore.Any())
-                await this.PushBlocksToRepositoryAsync(context);
-
-            this.logger.LogTrace("(-):{0}", StepResult.Stop);
-            return StepResult.Stop;
-        }
-
-        /// <summary>
-        /// When the node is in IBD wait for <see cref="BlockStoreLoop.PendingStorageBatchThreshold"/> to be true then continuously process all the blocks in <see cref="BlockStoreLoop.PendingStorage"/> until
-        /// a stop condition is found, the blocks will be pushed to the repository in batches of size <see cref="BlockStoreLoop.MaxPendingInsertBlockSize"/>.
-        /// </summary>
-        private async Task<StepResult> ProcessWhenInIBDAsync(ProcessPendingStorageContext context)
-        {
-            this.logger.LogTrace("({0}:'{1}',{2}.{3}:{4})", nameof(context.NextChainedBlock), context.NextChainedBlock, nameof(this.BlockStoreLoop.PendingStorage), nameof(this.BlockStoreLoop.PendingStorage.Count), this.BlockStoreLoop.PendingStorage.Count);
-
-            if (this.BlockStoreLoop.PendingStorage.Count < BlockStoreLoop.PendingStorageBatchThreshold)
-                return StepResult.Continue;
-
-            while (context.CancellationToken.IsCancellationRequested == false)
+            while (!context.CancellationToken.IsCancellationRequested)
             {
                 StepResult result = this.PrepareNextBlockFromPendingStorage(context);
                 if (result == StepResult.Stop)
                     break;
 
                 if (context.PendingStorageBatchSize > BlockStoreLoop.MaxPendingInsertBlockSize)
-                    await this.PushBlocksToRepositoryAsync(context);
-            }
-
-            if (context.PendingBlockPairsToStore.Any())
-                await this.PushBlocksToRepositoryAsync(context);
-
-            this.logger.LogTrace("(-):{0}", StepResult.Continue);
-            return StepResult.Continue;
-        }
-
-        /// <summary>
-        /// When the node is NOT in IBD, process and push the blocks in <see cref="BlockStoreLoop.PendingStorage"/> immediately
-        /// to the block repository without checking batch size.
-        /// </summary>
-        private async Task<StepResult> ProcessWhenNotInIBDAsync(ProcessPendingStorageContext context)
```

Comment:
@noescape00 Please check the logic here as when we are NOT in IBD we need to push to the repository immediately, without checking the batch size. That is why we had the separate methods as there are subtle differences in each scenario.

Your refactor caters for processing pending storage during IBD and disposing but not for "NotInIBD".

Reply:
here 
```
if (this.BlockStoreLoop.PendingStorage.Count < BlockStoreLoop.PendingStorageBatchThreshold &&
                !disposeMode && this.BlockStoreLoop.InitialBlockDownloadState.IsInitialBlockDownload())
            {
                return StepResult.Stop;
            }
```

we will never get into the returning stop if we're not in IBD so we will proceed to the loop that will remove all the blocks that we have in pending storage and push them immediately to the repository. 

Those lines: 

```
 if (context.PendingBlockPairsToStore.Any())
                await this.PushBlocksToRepositoryAsync(context).ConfigureAwait(false);
```

Will send the blocks to the repository no matter the batch size. 

So I don't see a problem here. 
Can you please elaborate?

Reply:
@noescape00 Actually, I just went through the logic again and the fact that we check the batch size in the while:

```
if (context.PendingStorageBatchSize > BlockStoreLoop.MaxPendingInsertBlockSize)
await this.PushBlocksToRepositoryAsync(context).ConfigureAwait(false);
```

Doesn't actually make a difference. So yeah, the logic is good 💯